### DISCUSSION TITLE
are startup events started when using pytests?

### DISCUSSION THREAD

**Discussion by nbraun-wolf** [fastapi/fastapi#4067] [Questions] [answered]
Hi, I feel like the startup tasks are not called when running pytests. I have created a small example of my problem, which you can find here. The issue is the pyests fail even though the app runs OK. 

https://github.com/nbraun-wolf/async-mqtt-fastapi-pytest

If someone has an idea why the pytests are not working and how to get them to work, it would be nice.

**Comment by ghandic** [comment]
They are, have you tried it with a minimal example, or dumbing your project down and slowly adding things back in?

**Reply by nbraun-wolf** [reply]
This is already a minimal example. I did precisely what you are asking, in that project. Did you look at it?

In another project, pytest coverage is telling me that this line isn't executed. It may be because it's a log statement, I am not sure, but it's another thing that makes me think startup tasks aren't executed.

```python
@root_app.on_event("startup")
def startup_log() -> None:
    get_logger().info(f"Listening on port {APP_PORT}...") # this one is a miss according to pytest coverage.
```

**Comment by Kludex** [comment, marked as answer]
httpx AsyncClient doesn't support lifespan events. Use asgi-lifespan alongside.

https://pypi.org/project/asgi-lifespan/

Check my project fastapi-microservices on the conftest you'll see an example. I'm busy right now, sorry.

**Reply by nbraun-wolf** [reply]
OK, I see. I will check it out and see if it helps. Thanks for the hint.

**Reply by Kludex** [reply]
https://github.com/Kludex/fastapi-microservices/blob/main/users/tests/conftest.py#L41-L44

**Reply by nbraun-wolf** [reply]
yes this runs the tasks. Thank you.
